src/main/java/com/minecolonies/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -828,9 +828,9 @@ private static ItemStack handleKeepX(@NotNull Map<ItemStorage, Integer> keptX,
      */
     private static boolean keptEnough(@NotNull Map<ItemStorage, Integer> kept, @NotNull Map<ItemStorage, Integer> keep, @NotNull ItemStack stack)
```

Comment:
Okay, so entry set and key set is a toss up at this point, because we are only ever actually getting one value from the map. In general if you need to iterate over a map, you shouldn't be calling map.get(). Instead you should be using the entry set, but since we only do one get, I'm not sure it matters. But I really think we should just re-evaluate the data structure we are using so that we don't have to iterate over a map. Iterating over a map is going to be a poor solution every time (If this is going to be impossible, maybe we should look at LinkedHashMap, which has better iterator performance). I think the biggest problem is that we are using ItemStorage as a key and want to do comparisons on several fields of ItemStorage.

```
keep.get(tempStorage) == null || (kept.get(tempStorage) != null && kept.get(tempStorage) >= keep.get(tempStorage))

 kept.get(tempStorage) >= keep.get(tempStorage) --> true always equal
(kept.get(tempStorage) != null && true) --> kept.get(tempStorage) != null

kept.get(tempStorage) == null || kept.get(tempStorage) != null --> true
```